core/api/src/main/java/com/microsoft/gctoolkit/GCToolKit.java

```diff
@@ -60,17 +51,40 @@ private static JavaVirtualMachine loadJavaVirtualMachine() {
     public GCToolKit() {
         // Allow for adding aggregations from source code,
         // but don't corrupt the ones loaded by the service loader
-        this.registeredAggregations = new HashSet<>(aggregationsFromServiceLoader);
+        this.registeredAggregations = new HashSet<>();
+    }
+
+    /**
+     * Loads all Aggregations defined in the application module through
+     * the java.util.ServiceLoader model. To register a class that
+     * provides the {@link Aggregation} API, define the following
+     * in {@code module-info.java}:
+     * <pre>
+     * import com.microsoft.gctoolkit.aggregator.Aggregation;
+     * import com.microsoft.gctoolkit.sample.aggregation.HeapOccupancyAfterCollectionSummary;
+     * 
+     * module com.microsoft.gctoolkit.sample {
+     *     ...
+     *     provides Aggregation with HeapOccupancyAfterCollectionSummary;
+     * }
+     * </pre>
+     */
+    public void loadAggregationsFromServiceLoader() {
+        if (aggregationsLoadedFromServiceLoader == true) return;
+
+        ServiceLoader.load(Aggregation.class)
+                .stream()
+                .map(ServiceLoader.Provider::get)
+                .map(Aggregation::getClass)
+                .forEach(registeredAggregations::add);
+
+        aggregationsLoadedFromServiceLoader = true;
```

Comment:
Is it possible that two threads run through this block at the same time? Then classes may be registered twice. so the whole block including boolean check should have a lock around it. The fact that it was in static {} before, made sure it was only executed once per loaded class.

Reply:
I believe my mistake here was letting the boolean as a static field. It shouldnt be. Would that clear your concern?

Reply:
Fixed in 4618237

Reply:
My concern was the fact that between the check and the assignment a context swap could be possible and a second threat could be waltzing in and execute the same logic also. The check-assignment approach is not a good lock. But since it is gone now, so is the concern.